information/specification.md

```diff
@@ -0,0 +1,706 @@
+    selection](#coin-selection) result.
+
+  * V<sub>_requested_</sub>
+
+    is the total value of the [requested output list](#requested-output-list)
+
+  * V<sub>_selected_</sub>
+
+    is the total value of the _inputs_ field of the [coin
+    selection](#coin-selection) result.
+
+### Conservation of UTxO
+
+This property states that the [initial UTxO set](#initial-utxo-set) is
+_conserved_ in the [results](#results).
+
+There are two mutually-exclusive cases:
+
+  * If a UTxO entry is _selected_ by the coin selection algorithm, it is
+    included in the _inputs_ field of the [coin selection](#coin-selection)
+    result.
+
+  * If a UTxO entry is _not selected_ by the coin selection algorithm, it is
+    included in the [remaining UTxO set](#remaining-utxo-set) result.
+
+The following laws hold:
+
+  * UTxO<sub>_selected_</sub> ⊆ UTxO<sub>_initial_</sub>
+  * UTxO<sub>_remaining_</sub> ⊂ UTxO<sub>_initial_</sub>
+  * UTxO<sub>_remaining_</sub> ∩ UTxO<sub>_selected_</sub> = ∅
+  * UTxO<sub>_remaining_</sub> ⋃ UTxO<sub>_selected_</sub> =
+    UTxO<sub>_initial_</sub>
+
+Where:
+
+  * **UTxO<sub>_initial_</sub>**
+
+    is the [initial UTxO set](#initial-utxo-set).
+
+  * **UTxO<sub>_remaining_</sub>**
+
+    is the [remaining UTxO set](#remaining-utxo-set).
+
+  * **UTxO<sub>_selected_</sub>**
+
+    is the value of the _inputs_ field of the [coin selection](#coin-selection)
+    result.
+
+### Conservation of Outputs
+
+This property states that the [requested output list](#requested-output-list)
+is _conserved_ in the [coin selection](#coin-selection) result.
+
+In particular, the _outputs_ field of the [coin selection](#coin-selection)
+result should be _equal to_ the [requested output list](#requested-output-list).
+
+### Cardinality
+
+All algorithms require that:
+
+ 1. Each output from the [requested output list](#requested-output-list) is
+    paid for by _one or more_ entries from the
+    [initial UTxO set](#initial-utxo-set).
+
+ 2. Each entry from the [initial UTxO set](#initial-utxo-set) is used to pay
+    for _at most one_ output from the [requested output
+    list](#requested-output-list).
+
+## Failure Modes
+
+There are a number of ways in which a coin selection algorithm can fail:
+
+  * #### UTxO Balance Insufficient
+
+    This failure occurs when the total value of the entries within the [initial
+    UTxO set](#initial-utxo-set) (the amount of money _available_) is _less
+    than_ the the total value of all entries in the [requested output
+    list](#requested-output-list) (the amount of money _required_).
+
+  * #### UTxO Not Fragmented Enough
+
+    This failure occurs when the _number_ of entries in the [initial UTxO
+    set](#initial-utxo-set) is _smaller than_ the number of entries in the
+    [requested output list](#requested-output-list).
+
+    All algorithms require that there is _at least one_ UTxO entry available
+    _for each_ requested output.
+
+  * #### UTxO Fully Depleted
```

Comment:
What is the difference between fully depleted and insufficient balance?

Reply:
@rvl wrote:
> What is the difference between fully depleted and insufficient balance?

Good question!

If the balance of a UTxO is insufficient to pay for the requested outputs, then that is a condition that we can detect _immediately_. It's equivalent to `sumValue utxo < sumValue outputs` (assuming a suitable `sumValue` function).

However, even if the balance _appears_ to be sufficient at the outset,  coin selection algorithms can _still_ end up exhausting all of the UTxO entries, due to the fact that coin selection algorithms don't always choose UTxO entries in a way that is "optimal". One complication here is the self-imposed restriction that any given UTxO entry can only be used to pay for a single output.

@KtorZ If we do run into the `UTxOFullyDepleted` condition (which I suspect is rare), is there a good reason to not relax the above restriction, so that, if needed, we allow a UTxO entry to pay for more than one output? This could be a very simple algorithm that doesn't attempt to perform any other optimization.

Reply:
@rvl I've made an additional change to clarify the difference here: 327efe4e07672933777a6b775463148e3817b1bc

Let me know if you think it can be clarified further!